src/Server.php

```diff
@@ -282,7 +282,9 @@ function ($error) use ($that, $conn, $request) {
             }
         );
 
-        if ($contentLength === 0) {
+        $upgradeConnection = $request->hasHeader('Connection') && $request->getHeaderLine('Connection') === 'Upgrade';
+
+        if (!$upgradeConnection && $contentLength === 0) {
             // If Body is empty or Content-Length is 0 and won't emit further data,
             // 'data' events from other streams won't be called anymore
             $stream->emit('end');
```

Comment:
This implies that the request body will *never* end for connection upgrade with an empty body.

Given that supporting the connection upgrades is optional as per the RFCs, this in turn implies that this creates potential for DOS attack if the consumer expects this `end` event.

For example, sending an upgrade request against example #04 will never return a result.

What do you think about this?

Afaict this is nothing that can be fixed with this design, so we *may* want to consider implementing #98 through a dedicated callback, as originally suggested?